### PR TITLE
Spectral Normalization continued

### PR DESCRIPTION
**Pull request by charlielito**
Continues the work of this abandoned PR https://github.com/tensorflow/addons/pull/376 which implements SpectralNormalization layer [1].

Changes:
* Fix `call` method for handling training and inference
* Added support for wrapping an embedding layer, since in [2] they used spectral normalization to the conditional embedding.
* Minor changes in raise error types and doc strings

Still to be done:
- [x] Finish tests

It would be great to get some feedback to finally finish this layer

[1] Miyato, T., Kataoka, T., Koyama, M., & Yoshida, Y. (2018). Spectral normalization for generative adversarial networks. arXiv preprint arXiv:1802.05957.
[2] Han Zhang, Ian Goodfellow, Dimitris Metaxas, Augustus Odena (2018). Self-Attention Generative Adversarial Networks. arXiv preprint arXiv:1805.08318

### CODE REVIEW

File: tensorflow_addons/layers/wrappers.py [hunk 1]
```diff
@@ -224,3 +224,112 @@ def remove(self):
             self.layer.kernel = kernel
 
         return self.layer
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class SpectralNormalization(tf.keras.layers.Wrapper):
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/wrappers.py]: I don't love expanding the wrappers module since most everything in Addons lives in it's own module. @gabrieldemarmiesse any thoughts on splitting this up to a spectral_normalization.py and weight_normalization.py?

File: tensorflow_addons/layers/wrappers.py [hunk 2]
```diff
@@ -224,3 +224,112 @@ def remove(self):
             self.layer.kernel = kernel
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/wrappers.py]: nit: Please update the copyright to 2020

File: tensorflow_addons/layers/wrappers_test.py [hunk 3]
```diff
@@ -174,5 +174,78 @@ def test_removal(self, base_layer_fn, input_shape):
...
+
+        self.assertTrue(hasattr(model.layers[0], "u"))
+
+    def test_applylayer(self):
+        images = tf.random.uniform((2, 4, 4, 3))
+        sn_wrapper = wrappers.SpectralNormalization(
+            tf.keras.layers.Conv2D(32, [2, 2]), input_shape=(4, 4, 3)
+        )
+        sn_wrapper.apply(images)
+        self.assertTrue(hasattr(sn_wrapper, "u"))
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/wrappers_test.py]: We'll want to verify these results against an expected value. May need to use a predetermined "image" or set a random seed.

File: tensorflow_addons/layers/wrappers_test.py [hunk 4]
```diff
@@ -174,5 +174,78 @@ def test_removal(self, base_layer_fn, input_shape):
...
+        input_data = np.random.random((10, 3, 4)).astype(np.float32)
+        test_utils.layer_test(
+            wrappers.SpectralNormalization,
+            kwargs={"layer": tf.keras.layers.Dense(2), "input_shape": (3, 4)},
+            input_data=input_data,
+        )
+
+
 if __name__ == "__main__":
     tf.test.main()
```
Comment (seanpmorgan) [inline tensorflow_addons/layers/wrappers_test.py]: Could you change this to `sys.exit(pytest.main([__file__]))` as per #1263. We're aiming to utilize pytest for its extensions.

Will require some imports

File: tensorflow_addons/layers/__init__.py [hunk 5]
```diff
@@ -24,3 +24,4 @@
 from tensorflow_addons.layers.sparsemax import Sparsemax
 from tensorflow_addons.layers.tlu import TLU
 from tensorflow_addons.layers.wrappers import WeightNormalization
+from tensorflow_addons.layers.wrappers import SpectralNormalization
```
[the remark this answers is not in the document]
Comment (charlielito) [reply tensorflow_addons/layers/__init__.py]: Resolved

File: tensorflow_addons/layers/wrappers.py [hunk 6]
```diff
@@ -224,3 +224,112 @@ def remove(self):
...
+    Raises:
+      AssertionError: If not initialized with a `Layer` instance.
+      ValueError: If initialized with negative `power_iterations`
+      AttributeError: If `Layer` does not contain a `kernel` or `embeddings` of weights
+    """
+
+    @typechecked
+    def __init__(self, layer: tf.keras.layers, power_iterations: int = 1, **kwargs):
+        super(SpectralNormalization, self).__init__(layer, **kwargs)
+        if power_iterations < 0:
```
[the remark this answers is not in the document]
Comment (charlielito) [reply tensorflow_addons/layers/wrappers.py]: You are right, it should run at least with one iteration. Already fixed this

File: tensorflow_addons/layers/wrappers.py (hunk 2 above)
Comment (charlielito) [reply tensorflow_addons/layers/wrappers.py]: Done

File: tensorflow_addons/layers/wrappers_test.py (hunk 3 above)
Comment (charlielito) [reply tensorflow_addons/layers/wrappers_test.py]: Seems reasonable, i'll finish tests and fix this

File: tensorflow_addons/layers/wrappers_test.py (hunk 4 above)
Comment (charlielito) [reply tensorflow_addons/layers/wrappers_test.py]: Updated

**Comment by Carrotor116** [comment]
I'm looking for spectral norm of tf 2 and then find this pull.

After checking the official implement which use func [`max_singular_value`](https://github.com/pfnet-research/sngan_projection/blob/master/source/functions/max_sv.py#L16) instead of `max_singular_value_fully_differentiable`, so `u` and `v` seem need to be apply stop_gradient after power iteration.

**Comment by Carrotor116** [comment]
In this implement, `layer.kernel` (`layer` is one which weights need to be apply spectral_norm) is update by `self.w.assign(self.w / sigma)` in function `normalize_weights`. It will block gradient from `normalize_weights` to _original_ `layer.kernel` . 

For example: 
```python 
import tensorflow as tf

def spectral_norm(w):
    return w / 2.  # no implement

x = tf.constant([[1, 2], [3, 4], [5, 6]], dtype=tf.float32)
kernel_val = [[1, 2, 3], [4, 5, 6]]

def foo1():
    kernel = tf.Variable(kernel_val, dtype=tf.float32)
    with tf.GradientTape() as tap:
        kernel_norm = spectral_norm(kernel)
        y = tf.matmul(kernel_norm, x)
    return y, tap.gradient(y, kernel)

def foo2():
    kernel = tf.Variable(kernel_val, dtype=tf.float32)
    with tf.GradientTape() as tap:
        kernel.assign(spectral_norm(kernel))
        y = tf.matmul(kernel, x)
    return y, tap.gradient(y, kernel)

if __name__ == '__main__':
    y1, grad1 = foo1()
    y2, grad2 = foo2()
    tf.assert_equal(y1, y2)
    tf.assert_equal(grad1, grad2)  # raise error

```

e.g.  applying `tf.assign` will clear the gradient from `spectral_norm`.
I thing `foo1` of the example should be a right way. In this way, replace `self.w.assign(self.w / sigma)` with `self.layer.kernel = self.w / sigma` :)

**Comment by charlielito** [comment]
@seanpmorgan Hey, sorry for the long wait. I've made some changes you proposed, and other tests. Also I took out the implementation to its own file, is that Ok as you suggested? I didn't do the suggestion of @Carrotor116  it crashes some of the tests.
One check is failing (the one that builds the wheel) I don't thing is related to this code. Could someone look into this?
Thanks!

**Comment by MokkeMeguru** [comment]
I think this pul request is great.
Won't you get up?

**Comment by charlielito** [comment]
> I think this pul request is great.
> Won't you get up?

I'm waiting for the code owners but they don't reply @seanpmorgan @facaiy

**Comment by wookayin** [comment]
Hi, thanks for your implementation. I found one bug in this implementation; this doesn't full work well with `@tf.function` being used.

Example:

```python
B = 10
model = networks.SpectralNormalization(tf.keras.layers.Dense(5))   
# model = tf.keras.layers.Dense(5)
model_fn = tf.function(model)
model_fn(tf.zeros([1, 3]))
model_fn(tf.zeros([5, 3]))
assert model.trainable_variables[0].shape == (3, 5)
```

This fails with an error (this runs OK on eager execution mode):

```
E         ValueError: in user code:                                                                
E                                                                                                  
E             .../lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py:975 __call__  *                                        
E                 input_spec.assert_input_compatibility(self.input_spec, inputs,                   
E             .../lib/python3.7/site-packages/tensorflow/python/keras/engine/input_spec.py:227 assert_input_compatibility  **                     
E                 ', found shape=' + str(shape))                                                   
E                                                                                                  
E             ValueError: Input 0 is incompatible with layer spectral_normalization: expected shape=[1, 3], found shape=[5, 3]                                                                         
```

File: tensorflow_addons/layers/spectral_normalization.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    def __init__(self, layer: tf.keras.layers, power_iterations: int = 1, **kwargs):
+        super(SpectralNormalization, self).__init__(layer, **kwargs)
+        if power_iterations <= 0:
+            raise ValueError(
+                "`power_iterations` should be greater than zero, got "
+                "`power_iterations={}`".format(power_iterations)
+            )
+        self.power_iterations = power_iterations
+        self._initialized = False
+        self._track_trackable(layer, name="layer")
```
Comment (WindQAQ) [inline tensorflow_addons/layers/spectral_normalization.py:58]: `Layer` is auto trackable, and has already been added to attr when calling `super`.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/wrappers.py#L53-L54

**Comment by WindQAQ** [comment]
> ```python
> model = networks.SpectralNormalization(tf.keras.layers.Dense(5))   
> # model = tf.keras.layers.Dense(5)       # <--- the normal dense layer works
> model_fn = tf.function(model)
> model_fn(tf.zeros([1, 3]))
> model_fn(tf.zeros([5, 3]))
> assert model.trainable_variables[0].shape == (3, 5)
> ```

Hi @wookayin, I think we can fix it by setting the first dim (batch dim) to `None` for `input_spec`.

https://colab.research.google.com/drive/10hL1gxrW3gKDh_i5s5ByThdpbHHAXS79?usp=sharing

**Comment by wookayin** [comment]
FYI: This implementation works very well for me https://github.com/google/edward2/blob/master/experimental/sngp/normalization.py

File: tensorflow_addons/layers/spectral_normalization.py (hunk 7 above)
Comment (charlielito) [reply tensorflow_addons/layers/spectral_normalization.py:58]: Sure

File: tensorflow_addons/layers/spectral_normalization.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    def build(self, input_shape):
+        """Build `Layer`"""
+
+        input_shape = tf.TensorShape(input_shape).as_list()
+        self.input_spec = tf.keras.layers.InputSpec(shape=input_shape)
+
+        if not self.layer.built:
+            self.layer.build(input_shape)
```
[the remark this answers is not in the document]
Comment (charlielito) [reply tensorflow_addons/layers/spectral_normalization.py:67]: Makes sense

**Comment by charlielito** [comment]
@WindQAQ suggestions were implemented

**Comment by charlielito** [comment]
@WindQAQ  the changes you suggested were implemented

File: tensorflow_addons/layers/tests/spectral_normalization_test.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    sn_layer = spectral_normalization.SpectralNormalization(base_layer)
+    model = tf.keras.models.Sequential(layers=[inputs, sn_layer])
+
+    # initialize model
+    model.predict(np.random.uniform(size=(2, 1)))
+
+    with tempfile.TemporaryDirectory() as model_dir:
+        model_path = str(Path(model_dir) / "saved_model")
+        model.save(model_path)
+        new_model = tf.keras.models.load_model(model_path)
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/spectral_normalization_test.py:59]: One minor thing. We can use pytest fixture here. We can change the function signature to
```python
def test_save_load_model(tmpdir):
```
The example usage can be found here
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/tests/kappa_loss_test.py#L95-L103

Comment (charlielito) [reply tensorflow_addons/layers/tests/spectral_normalization_test.py:59]: Nice, I didn't know about this

File: tensorflow_addons/layers/tests/spectral_normalization_test.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    input_shape = [1]
+
+    inputs = tf.keras.layers.Input(shape=input_shape)
+    sn_layer = spectral_normalization.SpectralNormalization(base_layer)
+    model = tf.keras.models.Sequential(layers=[inputs, sn_layer])
+
+    # initialize model
+    model.predict(np.random.uniform(size=(2, 1)))
+
+    model_path = str(Path(tmpdir) / "saved_model")
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/spectral_normalization_test.py:55]: Do not need to wrap it with `Path`.

File: tensorflow_addons/layers/tests/spectral_normalization_test.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    images = tf.ones((1, 2, 2, 1))
+    sn_wrapper = spectral_normalization.SpectralNormalization(
+        tf.keras.layers.Conv2D(
+            1, [2, 2], kernel_initializer=tf.constant_initializer(value=1)
+        ),
+        input_shape=(2, 2, 1),
+    )
+
+    result = sn_wrapper.__call__(images, training=False)
+    result_train = sn_wrapper.__call__(images, training=True)
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/spectral_normalization_test.py:150]: I think using `sn_wrapper(images, ...)` is enough?

File: tensorflow_addons/layers/wrappers.py [hunk 12]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
```
Comment (WindQAQ) [inline tensorflow_addons/layers/wrappers.py:1]: Please revert the change of this file

File: tensorflow_addons/layers/tests/spectral_normalization_test.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    input_shape = [1]
+
+    inputs = tf.keras.layers.Input(shape=input_shape)
+    sn_layer = spectral_normalization.SpectralNormalization(base_layer)
+    model = tf.keras.models.Sequential(layers=[inputs, sn_layer])
+
+    # initialize model
+    model.predict(np.random.uniform(size=(2, 1)))
+
+    model.save(tmpdir)
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/spectral_normalization_test.py:53]:
```suggestion
    model.save(tmpdir / "test.h5")
```

File: tensorflow_addons/layers/tests/spectral_normalization_test.py [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+    new_model = tf.keras.models.load_model(tmpdir)
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/spectral_normalization_test.py:54]:
```suggestion
    new_model = tf.keras.models.load_model(tmpdir / "test.h5")
```

File: tensorflow_addons/layers/tests/spectral_normalization_test.py [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+    inputs = tf.keras.layers.Input(shape=input_shape)
+    sn_layer = spectral_normalization.SpectralNormalization(base_layer)
+    model = tf.keras.models.Sequential(layers=[inputs, sn_layer])
+
+    # initialize model
+    model.predict(np.random.uniform(size=(2, 1)))
+
+    model.save(tmpdir / "test.h5")
+    new_model = tf.keras.models.load_model(tmpdir / "test.h5")
```
Comment (WindQAQ) [inline tensorflow_addons/layers/tests/spectral_normalization_test.py:54]: Hi @charlielito , sorry for the misleading info. Seems that keras does not accept `LocalPath` object. Let's try to wrap `tmpdir / "test.h5"` with `str`, that is, use `str(tmpdir / "test.h5")`.

Comment (charlielito) [reply tensorflow_addons/layers/tests/spectral_normalization_test.py]: Strangely, when I run it locally it works
